include/CL/cl_ext_pocl.h

```diff
@@ -60,42 +60,89 @@ typedef CL_API_ENTRY cl_int
 
 #endif
 
-/* cl_pocl_pinned_buffers (experimental stage) */
+/* cl_ext_buffer_device_address (experimental stage)
 
-#ifndef cl_pocl_pinned_buffers
-#define cl_pocl_pinned_buffers 1
-#define CL_POCL_PINNED_BUFFERS_EXTENSION_NAME "cl_exp_pinned_buffers"
+   TODO:
+   * Perhaps a better name could be simply 'cl_ext_device_pointers'?
+*/
+
+#ifndef cl_ext_buffer_device_address
+#define cl_ext_buffer_device_address 1
+
+/* We need both a platform and device extension due to the new buffer
+   creation flag. */
+
+/* clCreateBuffer(): A new cl_mem_flag CL_MEM_DEVICE_ADDRESS_EXT:
+
+   This flag specifies that the buffer must have a single fixed address
+   for its lifetime, which should be unique at least across the devices
+   of the context, but not necessarily with the host (virtual memory) as
+   with SVM or USM.
+
+   The flag might imply that the buffer will be "pinned" permanently to
+   a device's memory, but might not be necessarily so, as long as the address
+   range of the buffer remains constant.
+
+   The address is guaranteed to remain the same until the buffer is freed and
+   the address can be queried via clGetMemObjectInfo().
+
+   The device-specific buffer content updates must be still performed by
+   implicit or explicit buffer migrations performed by the runtime or the
+   client code. If any of the devices in the context does not support
+   such allocations, an error (CL_INVALID_VALUE) is returned.
+
+*/
+#define CL_MEM_DEVICE_ADDRESS_EXT (1ul << 31)
+
+/* Experimental multi-device allocation support flags. */
+
+/* If combined with CL_MEM_DEVICE_ADDRESS_EXT, each device can have their
+   own (fixed) device-side address and copy of the created buffer which are
+   synchronized implicitly. The main difference to a default cl_mem allocation
+   is then that the addresses are queriable with CL_MEM_DEVICE_PTRS_EXT and
+   the per-device address is guaranteed to be the same for the entire-lifetime
+   of the cl_mem.
 
-/* TODO: We need also platform/runtime extension due to the new buffer
-   creation flags. */
+   Otherwise, each compatible device in the context can access a shared
+   allocation in one of the device's global memory using the same address. */
+#define CL_MEM_DEVICE_PRIVATE_EXT (1ul << 30)
 
-/* clCreateBuffer(): A new cl_mem_flag CL_MEM_PINNED:
+/* clGetMemObjectInfo(): A new cl_mem_info type CL_MEM_DEVICE_PTR_EXT:
 
-   This flag specifies that the buffer must be persistently allocated
-   in the device's physical memory for its lifetime. That is, the buffer's
-   device address will remain the same and the space is reserved until
-   the buffer is freed. The device-specific buffer content updates are
-   still performed by implicit or explicit buffer migrations performed by
-   the runtime or the client code. If any of the devices in the context
-   does not support pinning, an error (TO DEFINE) is returned.
+   Returns the device address (as a void*) for a buffer allocated with
```

Comment:
I think returning `void*` is invalid here. Or at least needs some more comments. GPUs with a 64 bit VM can be used in 32 bit applications and we either require those pinned buffers to be allocated in the lower 32 bits or we relax that and make it a `cl_ulong`.

Reply:
Yes, it crossed my mind, but then I started to doubt if that's a realistic configuration these days or something worth trying to support. Is it? I'm not sure. Host void* allows 32b/32b (cl_ulong would force 32b systems to use 64b here), and 64b/32b systems to function, but the combination of 32b/64b host/device is indeed a corner case.

Reply:
It's hit by any 32 bit application on a modern system. However I doubt that applications still being 32 bit only would start using this extension.

But anyway, I think this would require a notice on the ext like "This requires a 64 bit host to function" or something like that. Or it simply moves to use `cl_ulong` as supporting this use case is generally pretty cheap. But I think `void*` should be changed to some int type anyway, so that applications don't assume it's a host accessible pointer or something.

In either case, I think it needs to be mentioned, so people don't ask later.

Reply:
I agree. Anyhow I think 32b/32b (where cl_ulong would be unnecessary) can be a valid ("deeply embedded") case. I'll go with a documentation fix for now and we can discuss more when we propose this to Khronos docs.